src/libertem/analysis/clust.py

```diff
@@ -24,47 +25,76 @@ class ClusterTemplate(GeneratorHelper):
+        ]
         return dep
 
     def temp_cluster_controller(self):
-        controller = inspect.getsource(ClusterAnalysis.controller)
-        # remove indentation
-        controller = controller.replace("\n        ", "\n")
-        controller = controller.split("\n\n")
-        temp_controller = []
-        controller[1] = controller[1].replace("self", "cluster_analysis")
-        controller[4] = controller[4].replace("self.", "")
-        controller[7] = controller[7].replace("self.dataset", "ds")
-        controller[2] = "sd_udf_results = ctx.run_udf(dataset=ds,\
-                         udf=stddev_udf, roi=roi, progress=True)"
-        temp_controller.extend(controller[1:9])
-        result = ["udf_results = ctx.run_udf(dataset=ds, udf=udf, progress=True)",
-                  "cluster_result = cluster_analysis.get_udf_results\
-                  (udf_results=udf_results, roi=roi)"]
-        temp_controller.append("\n".join(result))
-        return '\n\n'.join(temp_controller)
+        temp_controller = [
+                        "dataset = ds",
+                        "roi, sd_udf_results = get_sd_results()",
+                        "cluster_udf = get_cluster_udf(sd_udf_results)",
+                        "udf_results = ctx.run_udf(dataset=ds, udf=cluster_udf, progress=True)",
+                        "cluster_result = get_udf_results(udf_results, roi)",
+        ]
+        return '\n'.join(temp_controller)
 
     def get_docs(self):
         docs = ["# Cluster Analysis"]
         return '\n'.join(docs)
 
+    def get_run_sd_udf(self):
+        indent = " "*4
+        temp_run_sd = [
+            "def run_sd_udf(roi, stddev_udf):",
+            f"{indent}sd_udf_results = ctx.run_udf(dataset=ds, udf=stddev_udf, roi=roi,"
+            " progress=True)",
+            f"{indent}return roi, consolidate_result(sd_udf_results)"
+        ]
+        return "\n".join(temp_run_sd)
+
     def get_analysis(self):
-        temp_analysis = [f"parameters={self.params}",
-                         "cluster_analysis = ClusterAnalysis(dataset=ds, parameters=parameters)"]
-        temp_analysis.append(self.temp_cluster_controller())
-        return '\n'.join(temp_analysis)
+        udf_results = self.format_code(inspect.getsource(ClusterAnalysis.get_udf_results))
+        sd_roi = self.format_code(inspect.getsource(ClusterAnalysis.get_sd_roi))
+        cluster_udf = self.format_code(inspect.getsource(ClusterAnalysis.get_cluster_udf))
+        sd_results = self.format_code(inspect.getsource(ClusterAnalysis.get_sd_results))
+        run_sd_udf = self.get_run_sd_udf()
+        cluster_controller = self.temp_cluster_controller()
+
+        temp_analysis = [
+                    f"parameters={self.params}\ndataset=ds",
+                    f"{sd_roi}",
+                    f"{run_sd_udf}",
+                    f"{sd_results}",
+                    f"{cluster_udf}",
+                    f"{udf_results}",
```

Comment:
There should be vertical white space between the function definitions, ideally following PEP8 (two empty lines between top-level functions)

Reply:
There are currently two blank lines between the functions.

![image](https://user-images.githubusercontent.com/59721339/90312231-3d61f980-df20-11ea-927c-d1d2cea2a5ad.png)

Reply:
Hm, I didn't see the blank lines in the "copy notebook" view - maybe it is just an issue with the presentation in that view?